macaroonbakery/httpbakery/agent/agent.py

```diff
@@ -123,26 +127,22 @@ def legacy_interact(self, client, location, visit_url):
         '''Implement LegacyInteractor.legacy_interact by obtaining
         the discharge macaroon using the client's private key
         '''
+        log.info('agent legacy_interact, visit_url {}'.format(visit_url))
         agent = self._find_agent(location)
-        pk_encoded = self._auth_info.key.public_key.encode().decode('utf-8')
-        value = {
-            'username': agent.username,
-            'public_key': pk_encoded,
-        }
-        # TODO(rogpeppe) use client passed into interact method.
-        client = httpbakery.Client(key=self._auth_info.key)
-        client.cookies.set_cookie(utils.cookie(
+        client = copy.copy(client)
+        client.key = self._auth_info.key
+        resp = client.request(
+            method='POST',
             url=visit_url,
-            name='agent-login',
-            value=base64.urlsafe_b64encode(
-                json.dumps(value).encode('utf-8')).decode('utf-8'),
-        ))
-        resp = requests.get(url=visit_url, cookies=client.cookies,
-                            auth=client.auth())
+            json={
+                'username': agent.username,
+                'public_key': self._auth_info.key.public_key.encode().decode('utf-8'),
```

Comment:
.encode.decode? Are we changing the encoding here?

Reply:
encode is the method for serializing a PublicKey instance. Although the method name came from nacl, it seems that it's misleading. I've renamed "encode" to "serialize" and also added a __str__ method so we can avoid the stupid decode step.